flex-templates/java/regional_dlp_de_identification/src/main/java/org/apache/beam/samples/DLPTextToBigQueryStreaming.java

```diff
@@ -0,0 +1,780 @@
+              KV.of(
+                  c.element().getKey(),
+                  createBqRow(outputRow, headers.toArray(new String[headers.size()]))));
+        }
+      }
+    }
+
+    private static TableRow createBqRow(Table.Row tokenizedValue, String[] headers) {
+      TableRow bqRow = new TableRow();
+      AtomicInteger headerIndex = new AtomicInteger(0);
+      tokenizedValue
+          .getValuesList()
+          .forEach(
+              value -> {
+                String checkedHeaderName = checkHeaderName(headers[headerIndex.getAndIncrement()].toString());
+                bqRow.set(checkedHeaderName, value.getStringValue());
+              });
+      return bqRow;
+    }
+  }
+
+  private static String getJsonSchema(ValueProvider<String> inputSchema) {
+    String []fields = inputSchema.get().split(",");
+    String jsonSchema = "{ \"fields\": [";
+
+    for (int i = 0; i < fields.length; i++) {
+      jsonSchema += String.format("{\"name\": \"%s\", \"type\": \"%s\"}%s", fields[i].split(":")[0].trim(), fields[i].split(":")[1].trim(),
+          i == fields.length - 1 ? "": ",");
+    }
+    jsonSchema += "]}";
+    return jsonSchema;
+  }
+
+  private static String getFileName(ReadableFile file) {
+    String csvFileName = file.getMetadata().resourceId().getFilename().toString();
+    /** taking out .csv extension from file name e.g fileName.csv->fileName */
+    String[] fileKey = csvFileName.split("\\.", 2);
+
+    if (!fileKey[1].equals(ALLOWED_FILE_EXTENSION) || !TABLE_REGEXP.matcher(fileKey[0]).matches()) {
+      throw new RuntimeException(
+          "[Filename must contain a CSV extension "
+              + " BQ table name must contain only letters, numbers, or underscores ["
+              + fileKey[1]
+              + "], ["
+              + fileKey[0]
+              + "]");
+    }
+    /** returning file name without extension */
+    return fileKey[0];
+  }
+
+  private static BufferedReader getReader(ReadableFile csvFile) {
+    BufferedReader br = null;
+    ReadableByteChannel channel = null;
+    /** read the file and create buffered reader */
+    try {
+      channel = csvFile.openSeekable();
+
+    } catch (IOException e) {
+      LOG.error("Failed to Read File {}", e.getMessage());
+      throw new RuntimeException(e);
+    }
+
+    if (channel != null) {
+
+      br = new BufferedReader(Channels.newReader(channel, Charsets.UTF_8.name()));
+    }
+
+    return br;
+  }
+
+  private static List<String> getFileHeaders(BufferedReader reader) {
+    List<String> headers = new ArrayList<>();
+    try {
+      CSVRecord csvHeader = CSVFormat.DEFAULT.parse(reader).getRecords().get(0);
+      csvHeader.forEach(
+          headerValue -> {
+            headers.add(headerValue);
+          });
+    } catch (IOException e) {
+      LOG.error("Failed to get csv header values}", e.getMessage());
+      throw new RuntimeException(e);
+    }
+    return headers;
+  }
+
+  private static String checkHeaderName(String name) {
```

Comment:
consider if checking header names can be done earlier in the pipeline. I think it'll fail after DLP calls when doing translations.  could this be done after getting the csv headers, for instance.

Reply:
The checking and transformation, if necessary, of file headers is done in the "Process Tokenized Data" stage of the pipeline, which is just after the "DLP-Tokenization" stage, where data tokenization occurs by calling the DLP API.
 If the checking and transformation of the headers occurs at an earlier stage in the pipeline, the user will need to pass a DLP template with the headers already transformed, exactly as it is done in the code.
  For example, a header read from a csv file has the form "CVV/CVV2" and after the transformation is "CVVCVV2".
  In this way, the user would need to know exactly the transformation done in the code, instead of relying only on the headers in the input file, which the user has easy access to.
  Leaving it to the user to put into the DLP template the transformed headers as in the code, there is a high chance of errors. And since we are dealing with sensitive data here in the Data Warehouse repository, if the template does not have the headers exactly as they are transformed in the code, the data will not be de-identified and can be written into the BigQuery without this encryption.
  For these reasons, I think it is better to leave the checking and transformation of the headers where it is in the current code, what do you think @erlanderlo ?

Reply:
I opened an issue ( #296 ) for this.